### ISSUE TITLE
Exported PDF in Linux missing CJK character fonts in some platforms

### ISSUE THREAD

**Issue by taoky** [typst/typst#2032]
### Description

Version:

- typst 0.7.0 (da8367e1), from Arch official repo.

Reproduce:

```typst
= 这是标题 This is title
我能吞下玻璃而不伤身体 The quick fox jumps over the lazy dog.
```

And export to PDF with `typst compile`.

Okular in Linux and Microsoft Edge in Windows 10 displays this PDF correctly:

![image](https://github.com/typst/typst/assets/2109893/383ee152-385a-40f4-be44-062a3208faea)
![image](https://github.com/typst/typst/assets/2109893/83d91723-d069-4340-826d-66d994f9c20d)

But this is not the case in iOS:

![image](https://github.com/typst/typst/assets/2109893/45f9ab55-8473-41c9-a7cb-443dba9df54a)

And not in pdf.js (Firefox built-in PDF reader):

![image](https://github.com/typst/typst/assets/2109893/223713a3-76a5-4161-aab1-e45bfb6e79c9)

Output of `pdffonts`:

```
> pdffonts example.pdf
Syntax Warning: Mismatch between font type and embedded font file
Syntax Warning: Mismatch between font type and embedded font file
name                                 type              encoding         emb sub uni object ID
------------------------------------ ----------------- ---------------- --- --- --- ---------
ABCDEF+SourceHanSansCN-Bold          CID Type 0C (OT)  Identity-H       yes yes yes      3  0
ABCDEF+LinLibertineB                 CID TrueType      Identity-H       yes yes yes      8  0
ABCDEF+SourceHanSansCN-Regular       CID Type 0C (OT)  Identity-H       yes yes yes     13  0
ABCDEF+LinLibertine                  CID TrueType      Identity-H       yes yes yes     18  0
```

Compiled PDF file:
[example.pdf](https://github.com/typst/typst/files/12456098/example.pdf)

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
Possibly a duplicate of https://github.com/typst/typst/issues/573

**Comment by tiankaima** [comment]
> Possibly a duplicate of #573

(Collab on this issue), Just want to mention that when the exact same file is compiled on macOS, compiled PDF works just fine and CJK characters are shown without any trouble.

| Linux | macOS |
| --- | --- |
| <img width="1177" alt="image" src="https://github.com/typst/typst/assets/91816094/641f5a15-98bb-46f0-864d-3aeba3c0d7e7"> |  <img width="1177" alt="image" src="https://github.com/typst/typst/assets/91816094/4787bc1a-cfc3-41d4-8e24-ca3664f4af98"> |

Attached  [test.pdf](https://github.com/typst/typst/files/12458690/test.pdf) is compiled with the exact typst content:

```typst
= 这是标题 This is title
我能吞下玻璃而不伤身体 The quick fox jumps over the lazy dog.
```

**Comment by taoky** [comment]
@tiankaima It seems that this issue is related to fonts instead of specific OS. This issue and #573 are all related to Source Han fonts family (Source Han Sans in this issue, and Source Han Serif in #573). Your compiled PDF is using macOS's built-in Songti font so it works properly.

**Comment by taoky** [comment]
Some tests locally:

```typst
// Working properly with Noto CJK fonts (*.ttc)
// https://archlinux.org/packages/extra/any/noto-fonts-cjk/
// #set text(font: "Noto Sans CJK SC")
// #set text(font: "Noto Serif CJK SC")
// Not working with Source Han fonts (*.otf)
// https://archlinux.org/packages/extra/any/adobe-source-han-sans-cn-fonts/
// #set text(font: "Source Han Sans")
// https://archlinux.org/packages/extra/any/adobe-source-han-serif-cn-fonts/
// #set text(font: "Source Han Serif")
= 这是标题 This is title
我能吞下玻璃而不伤身体 The quick fox jumps over the lazy dog.
```

(You can use "Download from Mirror" link in Arch's package page to get the distributed font package.)

Is it a bug with OTF fonts...?

**Comment by tiankaima** [comment]
> Some tests locally:
> 
> ```
> // Working properly with Noto CJK fonts (*.ttc)
> // https://archlinux.org/packages/extra/any/noto-fonts-cjk/
> // #set text(font: "Noto Sans CJK SC")
> // #set text(font: "Noto Serif CJK SC")
> // Not working with Source Han fonts (*.otf)
> // https://archlinux.org/packages/extra/any/adobe-source-han-sans-cn-fonts/
> // #set text(font: "Source Han Sans")
> // https://archlinux.org/packages/extra/any/adobe-source-han-serif-cn-fonts/
> // #set text(font: "Source Han Serif")
> = 这是标题 This is title
> 我能吞下玻璃而不伤身体 The quick fox jumps over the lazy dog.
> ```
> 
> (You can use "Download from Mirror" link in Arch's package page to get the distributed font package.)
> 
> Is it a bug with OTF fonts...?

Building with ttc:

<img width="1177" alt="image" src="https://github.com/typst/typst/assets/91816094/061e37ba-875d-4ebd-83e4-e504286900ac">

**Comment by laurmaedje** [comment]
Might be fixed by [t/c0b332a4c07b95c0c3196d0d7b149560dcef67fa](https://github.com/typst/typst/commit/c0b332a4c07b95c0c3196d0d7b149560dcef67fa)

**Comment by laurmaedje** [comment]
Oh no. I will investigate whether it is because of the FIXME2 indicate in the code. If so, then at least I would know how to fix it.
